### PR TITLE
Allow the `CACHE_DIR` environment variable to override the cache location

### LINKED ISSUE #16: Controlling cache dir with env variable
### Issue
In many cases it is useful to control the cache directory depending on the environment some tool/command is run in. 

For example often in local development we want to keep the default `./node_modules/.cache/my-lib` location but in CI/CD pipeline we want to put it in some place we can persist it between runs.

It would be good to have an unified way of controlling directory that with env variable.

Env variables seem to be a common way to do that, a lot of projects that use caches support it in similar way (e.g. yarn uses `YARN_CACHE_FOLDER`, npm - `NPM_CACHE_LOCATION`).

#### Single variable for whole `.cache` dir
`CACHE_DIR=/some/path` will change the destination for all projects

Cons:
* no per-project granularity (might be useful to control e.g. what we're persisting between runs)
* `CACHE_DIR` is definitely a bad name (too general, 100% sure it's already used in many pipelines)
  * hard to name it in an understandable way

#### Separate variable for each project
When calling `findCacheDir({name: 'foo'})` we would use `FOO_CACHE_DIR`. 

Cons:
* name has to be generated (is `name.toUpperCase()` enough?)
* it will expose undocumented API to some libraries - might be confusing for library maintainers/developers who don't know about this feature

#### Every project does it separately
`find-cache-dir` doesn't handle the env variables, projects that want to use it handle it on their own.

Cons (?):
* has to be implemented manually on each project

> (Suggested on cache-loader: https://github.com/webpack-contrib/cache-loader/pull/92)

### PR DESCRIPTION
**Pull request by Richienb**
When the environmental variable `CACHE_DIR` is specified, the specified directory will be used instead. This is useful in development situations where packages write useful information to the cache.

Closes #16

### CODE REVIEW

**Comment by Richienb** [comment]
@sindresorhus

**Comment by sindresorhus** [comment]
Did you see my comment: https://github.com/avajs/find-cache-dir/issues/10#issuecomment-471871147 ?

**Comment by sindresorhus** [comment]
> This is useful in development situations where packages write useful information to the cache.

Can you be more specific about the use-case?

**Comment by Richienb** [comment]
@langri-sha Can you reference the packages you were talking about in https://github.com/avajs/find-cache-dir/issues/16#issuecomment-534279810?

**Comment by langri-sha** [comment]
In my project I have 8 dependencies that use `find-cache-dir`. Of them two are completely unconfigurable: `ava` and `@storybook/core`.

To save ~30s (~23$/mo), I have a local `find-cache-dir` package and use [Yarn's resolutions](https://yarnpkg.com/lang/en/docs/selective-version-resolutions/).

Feel free to use this if the PR doesn't get merged.

```
// package.json

{
  "resolutions": {
    "find-cache-dir": "file:your-workspace/find-cache-dir"
  }
}

// your-workspace/find-cache-dir/src/index.js

const findUp = require('find-up')
const path = require('path')
const { ensureDirSync } = require('fs-extra')

const root = findUp.sync('yarn.lock')

module.exports = (options = {}) => {
  const { name, create, thunk } = options
  
  const ensureDir = create ? ensureDirSync : path => path

  if (thunk) {
    return (...args) => ensureDir(path.join(root, '.cache', name, ...args))
  }

  return ensureDir(path.join('root', '.cache', name))
}
```

**Comment by Richienb** [comment]
@sindresorhus

**Comment by langri-sha** [comment]
I wish to correct myself on the point of `ava` as I was looking closely in the sources. It expresses the same behavior of `find-cache-dir`, but doesn't actually use it.

File: test.js [hunk 1]
```diff
@@ -20,6 +20,13 @@ test('finds from options.cwd', t => {
 	t.is(findCacheDir({cwd: __dirname, name: 'bar'}), path.join(__dirname, 'node_modules', '.cache', 'bar'));
 });
 
+test('supports CACHE_DIR environmental variable', t => {
+	const newCacheDir = path.join(__dirname, 'new-cache-dir');
```
Comment (sindresorhus) [inline test.js]: Use https://github.com/sindresorhus/tempy to put it in a temporary location.

**Comment by sindresorhus** [comment]
`CACHE_DIR` is a very generic environment variable, which is fine, but I think we should not put the cache directly in `CACHE_DIR`, but rather inside `$CACHE_DIR/find-cache-dir`, to prevent possible conflicts with other projects using that directory.

File: index.js [hunk 2]
```diff
@@ -15,32 +15,39 @@ const isWritable = path => {
...
-		if (directory && options.create) {
-			makeDir.sync(directory);
-		}
+			directory = path.join(directory, 'node_modules', '.cache', name);
 
-		if (options.thunk) {
-			return (...arguments_) => path.join(directory, ...arguments_);
+			if (directory) {
+				directory = path.join(directory, 'node_modules', '.cache', name);
+			}
```
Comment (sindresorhus) [inline index.js]: This logic makes no sense to me.

File: index.js [hunk 3]
```diff
@@ -15,32 +15,39 @@ const isWritable = path => {
...
 		}
 	}
```
Comment (sindresorhus) [inline index.js]: ?

File: readme.md [hunk 4]
```diff
@@ -45,7 +45,7 @@ findCacheDir({name: 'unicorns'});
 
 ### findCacheDir(options?)
 
-Finds the cache directory using the supplied options. The algorithm tries to find a `package.json` file, searching every parent directory of the `cwd` specified (or implied from other options). It returns a `string` containing the absolute path to the cache directory, or `undefined` if `package.json` was never found or if the `node_modules` directory is unwritable.
+Finds the cache directory using the supplied options. The algorithm checks for the CACHE_DIR environmental variable and if one is not found, it tries to find a `package.json` file, searching every parent directory of the `cwd` specified (or implied from other options). It returns a `string` containing the absolute path to the cache directory, or `undefined` if `package.json` was never found or if the `node_modules` directory is unwritable.
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
Finds the cache directory using the supplied options. The algorithm checks for the `CACHE_DIR` environmental variable, and if one is not found, it tries to find a `package.json` file, searching every parent directory of the `cwd` specified (or implied from other options). It returns a `string` containing the absolute path to the cache directory, or `undefined` if `package.json` was never found or if the `node_modules` directory is unwritable.
```

File: test.js [hunk 5]
```diff
@@ -20,6 +20,13 @@ test('finds from options.cwd', t => {
...
+	process.env.CACHE_DIR = newCacheDir;
+	t.is(findCacheDir(), newCacheDir);
+	delete process.env.CACHE_DIR;
+});
```
Comment (sindresorhus) [inline test.js]: Would be nice with more than just one test.

File: readme.md [hunk 6]
```diff
@@ -45,7 +45,7 @@ findCacheDir({name: 'unicorns'});
 
 ### findCacheDir(options?)
```
Comment (sindresorhus) [inline readme.md]: The readme also needs to mention some of the use-cases for `CACHE_DIR`.

File: test.js (hunk 1 above)
Comment (Richienb) [reply test.js]: Done!

File: index.js (hunk 2 above)
Comment (Richienb) [reply index.js]: I rewrote the logic based on the master branch copy.

File: test.js (hunk 5 above)
Comment (Richienb) [reply test.js]: Tests added!

File: readme.md (hunk 6 above)
Comment (Richienb) [reply readme.md]: Done!

**Comment by Richienb** [comment]
@sindresorhus Ready for re-review!

**Comment by sindresorhus** [comment]
The code looks technically correct, but I feel like it could be refactored a bit to improve readability.

https://github.com/Richienb/find-cache-dir/blob/ea4a195f23faee1f150e7342dd136aefdb603c54/index.js#L20 is kinda awkward, if true, the directory is the actual cache directory, if false, it's the CWD directory.

This could be made a separate method: https://github.com/Richienb/find-cache-dir/blob/ea4a195f23faee1f150e7342dd136aefdb603c54/index.js#L32-L40 which could also include this part: https://github.com/avajs/find-cache-dir/blob/8ec64dcb35b3975e13a08dba2a3b6e7fb599ae30/index.js#L19-L25

That way the main method logic only handles the resolved cache directories.

File: test.js [hunk 7]
```diff
@@ -1,8 +1,9 @@
 import fs from 'fs';
 import path from 'path';
+import findCacheDir from '.';
 import {serial as test} from 'ava';
 import del from 'del';
-import findCacheDir from '.';
+import {directory as tempDir} from 'tempy';
```
Comment (sindresorhus) [inline test.js]:
```suggestion
import {directory as tempDirectory} from 'tempy';
```

File: test.js [hunk 8]
```diff
@@ -39,3 +40,20 @@ test('returns undefined if it can\'t find package.json', t => {
 	process.chdir(path.join(__dirname, '..'));
 	t.is(findCacheDir({name: 'foo'}), undefined);
 });
+
+test('supports CACHE_DIR environmental variable', t => {
```
Comment (sindresorhus) [inline test.js]:
```suggestion
test('supports CACHE_DIR environment variable', t => {
```

File: test.js [hunk 9]
```diff
@@ -39,3 +40,20 @@ test('returns undefined if it can\'t find package.json', t => {
...
+	const newCacheDir = tempDir();
```
Comment (sindresorhus) [inline test.js]:
```suggestion
	const newCacheDirectory = tempDir();
```

**Comment by sindresorhus** [comment]
I think it could still be made more readable. And this is not resolved:

> https://github.com/Richienb/find-cache-dir/blob/ea4a195f23faee1f150e7342dd136aefdb603c54/index.js#L20 is kinda awkward, if true, the directory is the actual cache directory, if false, it's the CWD directory.

**Comment by Richienb** [comment]
@sindresorhus I found a refactoring strategy to make the code more readable.

File: index.js [hunk 10]
```diff
@@ -14,8 +16,36 @@ const isWritable = path => {
...
+}
+
+function getNodeModuleDirectory(directory) {
+	const nodeModules = path.join(directory, 'node_modules');
+
+	if (
+		!isWritable(nodeModules) &&
+		(fs.existsSync(nodeModules) || !isWritable(path.join(directory)))
+	) {
+		return undefined;
```
Comment (sindresorhus) [inline index.js]: It's moot to return `undefined`. An empty return is the same.

**Comment by sindresorhus** [comment]
Looks good :)

**Pull request merged**